### PR TITLE
[SPARK-10439] [sql] Add bound checks to DateTimeUtils.

### PR DESCRIPTION
**Pull request by vanzin**
There were a couple of places where Spark SQL would silently truncate
data if certain timestamps were provided.

In a couple of other places, the way to calculate Julian day-based
timestamps was changed a little so that Spark writes data that is
friendlier to Hive; mostly, Hive does not like very much when
the data has negative values for either the days or nanos part,
so avoid those.

The values that trigger these use cases are very uncommon (very large
values in either end of the spectrum), so this shouldn't really affect
any existing applications.

### CODE REVIEW

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
@yhuai I resolved a conflict cause by #8597. In that PR you added bounds to the timestamps generated by `RandomDataGenerator`, but there's no real explanation of why you chose those values.

I took those at face value, so now all timestamps are checked against those bounds, even though they are more restrictive than the ones I had previously (e.g., now you can't represent any BC timestamps). It would be nice to have a better explanation for why those bounds are there, though.

**Comment by yhuai** [comment]
@vanzin I chose those values just to make the random data generator generates valid data for tests. Feel free to change them to more reasonable values.

**Comment by vanzin** [comment]
ok, thanks, I'll revert to the bounds I had before. I was assuming you picked those values because of some problem you ran while testing that change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -56,6 +57,12 @@ object DateTimeUtils {
 
   @transient lazy val defaultTimeZone = TimeZone.getDefault
 
+  // Constants defining the allowed ranges for timestampts that fit in parquet files. Mostly
+  // used by tests to avoid duplication.
+  private[spark] final val MIN_TIMESTAMP: Long =
+    -(DateTimeUtils.JULIAN_DAY_OF_EPOCH * DateTimeUtils.SECONDS_PER_DAY * 1000)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I gave it a try and I got 
```
scala> new java.sql.Timestamp(-4708372992000L)
res3: java.sql.Timestamp = 1820-10-18 14:36:48.0

scala> java.sql.Timestamp.valueOf("0317-02-14 06:13:20.0")
res7: java.sql.Timestamp = 0317-02-14 06:13:20.0

scala> res5.getTime
res8: Long = -52159715200000

scala> new java.sql.Timestamp(-72135740800000L)
res1: java.sql.Timestamp = 0317-02-14 06:13:20.0
```

So, is it the right lower bound?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -56,6 +57,12 @@ object DateTimeUtils {
...
+  private[spark] final val MAX_TIMESTAMP: Long = java.lang.Long.MAX_VALUE / 1000
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Also, although we can do 
```
scala> new java.sql.Timestamp(java.lang.Long.MAX_VALUE / 1000)
res13: java.sql.Timestamp = 294247-01-09 20:00:54.775
```

but, we cannot do `java.sql.Timestamp.valueOf("294247-01-09 20:00:54.775")`. There is 
```
java.lang.IllegalArgumentException: Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]
	at java.sql.Timestamp.valueOf(Timestamp.java:235)
	at .<init>(<console>:8)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:734)
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:983)
	at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:573)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:604)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:568)
	at scala.tools.nsc.interpreter.ILoop.reallyInterpret$1(ILoop.scala:760)
	at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:805)
	at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:717)
	at scala.tools.nsc.interpreter.ILoop.processLine$1(ILoop.scala:581)
	at scala.tools.nsc.interpreter.ILoop.innerLoop$1(ILoop.scala:588)
	at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:591)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:882)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:837)
	at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:837)
	at scala.tools.nsc.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:135)
	at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:837)
	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:83)
	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:96)
	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:105)
	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```
For text based format, the upper bound of timestamp value is much smaller.

**Comment by yhuai** [comment]
also cc @davies since he is a better person to review this one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 1 above)
Comment (vanzin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: It's the right value if you consider the math. But I've seen really weird behavior in how the Java classes print very large (positive or negative) timestamps, and I didn't find anything in the javadocs about limits, so I'm not sure what's the best way to proceed. We can choose an arbitrary minimum and maximum, but what would those be based on?

For example, `java.sql.Timestamp` will print the same formatted date string for both `-793495812000` and `-123456789012000L`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 2 above)
Comment (vanzin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Yes, the ranges might be different for the string parser, but is there any documentation about what that range would be?

**Comment by vanzin** [comment]
@yhuai I had to revert to the limits you had before because of the test failures; at least that helped me provide a better comment for the limits, instead of just magic numbers. :-p

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 3]
```diff
@@ -82,7 +90,9 @@ object DateTimeUtils {
     // SPARK-6785: use Math.floor so negative number of days (dates before 1970)
     // will correctly work as input for function toJavaDate(Int)
     val millisLocal = millisUtc + threadLocalLocalTimeZone.get().getOffset(millisUtc)
-    Math.floor(millisLocal.toDouble / MILLIS_PER_DAY).toInt
+    val days = Math.floor(millisLocal.toDouble / MILLIS_PER_DAY)
+    require(days <= Integer.MAX_VALUE && days >= Integer.MIN_VALUE, "Date exceeeds allowed range.")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I'd assume the millisUtc is a valid timestamp, then this check is not needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 4]
```diff
@@ -193,9 +205,15 @@ object DateTimeUtils {
    */
   def toJulianDay(us: SQLTimestamp): (Int, Long) = {
     val seconds = us / MICROS_PER_SECOND
-    val day = seconds / SECONDS_PER_DAY + JULIAN_DAY_OF_EPOCH
-    val secondsInDay = seconds % SECONDS_PER_DAY
-    val nanos = (us % MICROS_PER_SECOND) * 1000L
+    var day = seconds / SECONDS_PER_DAY + JULIAN_DAY_OF_EPOCH
+    var secondsInDay = seconds % SECONDS_PER_DAY
+    var nanos = (us % MICROS_PER_SECOND) * 1000L
+    if (us < 0) {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Good catch!

**Comment by davies** [comment]
I'm not a super fan to add bound checking to those heavily used internal functions. If we really want to limit the range that we support, we should do in inbound face (for data sources or in CatalystConverter) and after calculation.

**Comment by vanzin** [comment]
@davies sorry, not sure I follow.

`millisToDays` seems to not need the check, so I'll remove it. I didn't find any code path that started from user code. But `fromJavaTimestamp` is called from user code - the you convert an `RDD[Row]` to a `DataFrame`. I think it's better to fail early in that case.

The check in `toJulianDay` should be rarely invoked, and might be even unnecessary given the current limits.

**Comment by vanzin** [comment]
If you think it would be better I could move the checks to `CatalystTypeConverters` (DateConverter - which is missing from my patch! - and TimestampConverter).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 5]
```diff
@@ -193,9 +203,14 @@ object DateTimeUtils {
...
+      day -= 1
+      secondsInDay += SECONDS_PER_DAY
+      nanos += (SECONDS_PER_DAY * MICROS_PER_SECOND * 1000L)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: This is not correct, it should be:
```

val julian_us = us + JULIAN_DAY_OF_EPOCH * MICROSECONDS_PER_DAY
val day = julian_us / MICROSECONDS_PER_DAY
val micros = julian_us % MICROSECONDS_PER_DAY
(day.toInt, micros * 1000L
```

Comment (vanzin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Good catch; my code can return `nanos > 999999999` which `java.sql.Timestamp` does not like...

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Created https://issues.apache.org/jira/browse/SPARK-10522 for this, then we could back port the fix for 1.5.

**Pull request closed without merging**